kong/db/dao/certificates.lua

```diff
@@ -0,0 +1,211 @@
+local singletons = require "kong.singletons"
+local cjson      = require "cjson"
+local utils      = require "kong.tools.utils"
+
+-- Get an array of SNI names from either a string (split+sort),
+-- an array(sort) or ngx.null(return {})
+-- Returns an error if the list has duplicates
+-- Returns nil if input is falsy.
+local function parse_name_list(input, errors)
+  local name_list
+  if type(input) == "string" then
+    name_list = utils.split(input, ",")
+
+  elseif type(input) == "table" then
+    name_list = utils.shallow_copy(input)
+
+  elseif input == ngx.null then
+    name_list = {}
+  end
+
+  if not name_list then
+    return nil
+  end
```

Comment:
The code would be simplified if we avoided returning `nil` from this function, and instead always return a table (empty if no names):

```lua
if type(input) == "table" then
  name_list = utils.shallow_copy(input)

elseif input == ngx.null then
  name_list = {}

else
  -- maybe throw an error here, looks like a programming error
end

-- ...

return setmetatable(name_list, cjson.empty_array_mt)
```

This way, the below `insert_with_name_list()` method could do: 
```lua
if #name_list > 0 then

end

-- ...

cert.snis = name_list -- no need for cjson.empty_array fallback
```

Reply:
The idea I was going for was this:

* If you don't set the `snis` variable at all, it remains as it was (no programming error, maybe you where just missing a char on the key and are updating just that).
* If you set the `snis` variable, it will create/delete whatever snis is necessary. This includes passing an empty table (all snis will be removed)
* If you pass a json null, it is as if you passed an empty table.

If I do the change you propose, there are two drawbacks:
* Either we consider that an empty table means "no changes" (which I think is confusing).
* Or we require the user to pass the `snis` param every time they want to make a change on a cert, even for changing the cert or key (otherwise we set of an error).

If you think this is worth it, I will implement it, but I wanted to warn you about the costs first.